src/audio/rtnr/rtnr.c

```diff
@@ -0,0 +1,549 @@
+	return dev;
+
+cd_fail:
+	comp_data_blob_handler_free(cd->model_handler);
+	rfree(cd);
+
+err:
+	rfree(dev);
+	return NULL;
+}
+
+static void rtnr_free(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	comp_info(dev, "rtnr_free()");
+
+	comp_data_blob_handler_free(cd->model_handler);
+
+	RTKMA_API_Context_Free((void *)&cd->rtk_agl);
+	free_RtkMemoryPool();
+
+	rfree(cd);
+	rfree(dev);
+}
+
+/* set component audio stream parameters */
+static int rtnr_params(struct comp_dev *dev, struct sof_ipc_stream_params *params)
+{
+	int ret;
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *sinkb;
+	struct comp_buffer *sourceb;
+
+	comp_info(dev, "rtnr_params()");
+
+	ret = comp_verify_params(dev, 0, params);
+	if (ret < 0) {
+		comp_err(dev, "rtnr_params() error: comp_verify_params() failed.");
+		return ret;
+	}
+
+	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer,
+				  sink_list);
+	sinkb = list_first_item(&dev->bsink_list, struct comp_buffer,
+				source_list);
+
+	/* set source/sink_frames/rate */
+	cd->source_rate = sourceb->stream.rate;
+	cd->sink_rate = sinkb->stream.rate;
+	if (!cd->sink_rate) {
+		comp_err(dev, "rtnr_nr_params(), zero sink rate");
+		return -EINVAL;
+	}
+
+	/* Currently support 16kHz sample rate only. */
+	switch (sourceb->stream.rate) {
+	case 16000:
+		comp_info(dev, "rtnr_params(), sample rate = 16000 kHz");
+		break;
+	default:
+		comp_err(dev, "rtnr_nr_params(), invalid sample rate(%d kHz)",
+				 sourceb->stream.rate);
+		return -EINVAL;
+	}
+
+	if (sourceb->stream.channels != sinkb->stream.channels) {
+		comp_err(dev, "rtnr_params(), mismatch source/sink stream channels");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int rtnr_cmd_get_data(struct comp_dev *dev,
+			    struct sof_ipc_ctrl_data *cdata, int max_size)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		comp_info(dev, "rtnr_cmd_get_data(), SOF_CTRL_CMD_BINARY");
+		ret = comp_data_blob_get_cmd(cd->model_handler, cdata, max_size);
+		break;
+	default:
+		comp_err(dev, "rtnr_cmd_get_data() error: invalid command %d", cdata->cmd);
+		ret = -EINVAL;
+		break;
+	}
+	return ret;
+}
+
+static int rtnr_cmd_set_data(struct comp_dev *dev,
+			    struct sof_ipc_ctrl_data *cdata)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		comp_info(dev, "rtnr_cmd_set_data(), SOF_CTRL_CMD_BINARY");
+		ret = comp_data_blob_set_cmd(cd->model_handler, cdata);
+		break;
+	default:
+		comp_err(dev, "rtnr_cmd_set_data() error: invalid command %d", cdata->cmd);
+		ret = -EINVAL;
+		break;
+	}
+	return ret;
+}
+
+static int32_t rtnr_cmd_get_value(struct comp_dev *dev, struct sof_ipc_ctrl_data *cdata)
+{
+	int32_t ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_SWITCH:
+		comp_dbg(dev, "rtnr_cmd_get_value(), SOF_CTRL_CMD_SWITCH, cdata->comp_id = %u",
+			 cdata->comp_id);
```

Comment:
if not implemented should return `-EINVAL`

Reply:
The set/get_value are implemented. Thanks.